PICO split and merge

talbrecht opened the pull request:
This pull request adds some special treatment of ice shelves and ocean inputs in the **PICO** ocean module (e.g. in case of a collapse of the West Antarctic Ice Sheet). When ice shelves merge across basin boundaries that both have access to the ocean, but that do not share a common ocean boundary (no basin neighbors at the coast), then those ice shelves are split into two separate instances with individual ocean inputs. When a grounding line retreats into another basins, which does not has a connection to the ocean in that basin (ice shelf cavity), ocean conditions from this basin are not considered in the area-weighted average of the ocean inputs for this ice shelf instance.

**Checklist**

- [x] update documentation
- [x] update [`CHANGES.rst`](CHANGES.rst)
- [x] add regression tests

ckhroulev commented on src/coupler/ocean/PicoGeometry.cc:
```diff
@@ -468,6 +501,167 @@ void PicoGeometry::compute_ocean_mask(const IceModelVec2CellType &cell_type, Ice
+      }
+    }
+
+    //GlobalSum(m_grid->com, cfs_in_basins_per_shelf.data(), m_n_shelves*m_n_basins);
+    //GlobalSum(m_grid->com, n_shelf_cells_per_basin.data(), m_n_shelves*m_n_basins);
```
Please remove commented out code to reduce clutter.

ckhroulev commented on src/coupler/ocean/PicoGeometry.cc:
```diff
@@ -468,6 +501,167 @@ void PicoGeometry::compute_ocean_mask(const IceModelVec2CellType &cell_type, Ice
+    }
+  }
+
+  //GlobalSum(m_grid->com, n_shelf_cells_to_split.data(), m_n_shelves*m_n_basins);
```
Please remove commented out code to reduce clutter.

ckhroulev commented on test/test_pico_split/run_test.sh:
```diff
@@ -0,0 +1,106 @@
+
+# create mean salinity and ocean temperature fields
+cp $INNAME input_test01.nc
+#ncap2 -O -s "salinity_ocean=0*topg;theta_ocean=0*topg" input_test01.nc input_test01.nc
+#ncatted -O -a units,salinity_ocean,o,c,"g/kg" \
+#           -a cell_method,salinity_ocean,o,c,"time: mean" \
+#           -a long_name,salinity_ocean,d,, \
+#           -a pism_intent,salinity_ocean,d,, \
+#           -a standard_name,salinity_ocean,d,, \
+#           -a units,theta_ocean,o,c,"Celsius" \
+#           -a cell_method,theta_ocean,o,c,"time: mean" \
+#           -a long_name,theta_ocean,d,, \
+#           -a pism_intent,theta_ocean,d,, \
+#           -a standard_name,theta_ocean,d,, \
+#           input_test01.nc
+
+# fill in mean ocean values 
+#bi=0
+#for ti in "${tlist[@]}"
+#do
+#  ncap2 -O -s "where(basins==${bi}) theta_ocean=${ti}" input_test01.nc input_test01.nc
+#  bi=$((bi+1))
+#bi=0
+#for si in "${slist[@]}"
+#do
+#  ncap2 -O -s "where(basins==${bi}) salinity_ocean=${si}" input_test01.nc input_test01.nc
+#  bi=$((bi+1))
+#done
+
+#ncap2 -O -s "climatic_mass_balance=0*salinity_ocean" input_test01.nc input_test01.nc
+#ncatted -O -a units,climatic_mass_balance,o,c,"kg m-2 year-1" \
+#           -a long_name,climatic_mass_balance,o,c,"surface mass balance (accumulation/ablation) rate" \
+#           -a standard_name,climatic_mass_balance,o,c,"land_ice_surface_specific_mass_balance_flux" \
+#           input_test01.nc
```
Here and below: please remove commented out code to reduce clutter.

ckhroulev commented on src/coupler/ocean/Pico.cc:
```diff
@@ -414,12 +420,33 @@ void Pico::set_ocean_input_fields(const PicoPhysics &physics, const IceModelVec2
         n_shelf_cells_per_basin[s][b] = GlobalSum(m_grid->com, n_shelf_cells_per_basin[s][b]);
+
+        cfs_in_basins_per_shelf[s][b] = GlobalSum(m_grid->com, cfs_in_basins_per_shelf[s][b]);
+        //if (cfs_in_basins_per_shelf[s][b]>0) {
+        //  m_log->message(2, "PICO, ocean gateways results with s=%d, r=%d \n",s,cfs_in_basins_per_shelf[s][b]);
+        //}
```
Please remove commented out code to reduce clutter.

talbrecht replied:
Oops, I guess I rebased merge_pico instead of pico_merge... Will update this one later...

ckhroulev reviewed (changes requested):
Looks good, but needs a bit of tidying up. Thanks!

talbrecht commented:
OK, pik/pico_merge is now rebased to dev, cleaned up old comments, and GlobalSums reformulated for better performance. The test_pico_split has been updated...

ckhroulev commented:
The current version looks good to me. @talbrecht, please run a couple of tests to check that it still works the way you intended.

And one more thing: we need to automate the regression test. The shell script should return 0 on success and something else (1, for example) on failure (just add `exit 0` and `exit 1` in the right places).

talbrecht commented:
A added a lot of `if` and `else`, but it seems to work as a quick regression test... I guess the `echo`s could be removed then?

ckhroulev commented:
> I guess the echos could be removed then?

They aren't hurting anything, so I'd leave it like this.

Thanks!

talbrecht commented:
The regression test works fine... Always good to learn some more bash...
